hips/hip-00NN.md

```diff
@@ -0,0 +1,101 @@
+- testing work-in-progress charts in their umbrella charts
+
+## Motivation
+
+There are existing ways to achieve installation of charts without using a registry, however they are not very user-friendly and require additional tooling.
+
+- A Helm chart repository is effectively an `index.yaml` with links to downloads. Maintaining such a repository does create a burden of scripts and automation (e.g. Github Actions). This is not always feasible for smaller projects. It also does not really offer an obvious and readily available way of testing pre-releases.
+- For the testing use cases, charts can be packaged using `helm package`, however this does introduce manual steps and requires extra work to replicate in CI/CD scenarios.
+- There are several plugins available to solve this problem ([`aslafy-z/helm-git`](https://github.com/aslafy-z/helm-git), [`diwakar-s-maurya/helm-git`](https://github.com/diwakar-s-maurya/helm-git), [`sagansystems/helm-github`](https://github.com/sagansystems/helm-github)), however using plugins requires additional setup, which may not always be feasible (esp. in more complex team structures and cluster setups with advanced tooling, e.g. ArgoCD). An additional drawback is that the `Chart.yaml` does not provide a way to specify the plugin requirements, which leaves it up to the consumer to figure this out.
+
+## Rationale
+
+Installing dependencies from git is an established pattern in other ecosystems even when they are registry-based, e.g. npm (Node.js), pipenv (Python), bundler (Gem) have this option - it would make sense to have the behavior replicated in Helm.
+
+At least the npm and the pip ecosystems have already established a syntax of `vcs+protocol` for defining the dependency source, so it should be familiar to some users.
+
+One alternative to consider would be to exclude defining the vcs/protocol, similar to what Go does, esp. given that Helm is built using Go. This does limit the flexibility somewhat - while Go does allow adding a VCS qualifier at the end of the URL (allowing future support for other VCSs), it does not allow specifying the protocol, which means that the users might have to override the default protocol in their VCS configuration.
+
+## Specification
+
+The `Chart.yaml` should support the following format for `dependencies`:
+
+```
+dependencies:
+- name: "<dependency name>"
+  repository: "<protocol>://[<user>[:<password>]@]<hostname>[:<port>][:][/]<path>"
+  version: "<commit-ish>"
+```
+where:
+- `<protocol>` is one of `git`, `git+ssh`, `git+http`, `git+https`, or `git+file`.
+- `<commit-ish>` is an existing reference (SHA hash, tag or branch name) on the repo.
+
+For example:
+
+```
+dependencies:
+- name: "jenkins"
+  repository: "git+https://github.com/jenkinsci/helm-charts.git/charts/jenkins"
+  version: "main"
+```
+
+When Helm is installing a dependency from git, it should:
+
+- create a temporary directory
```

Comment:
Is this the process that other similar systems follow? (go modules, npm, pipenv, etc. And the mentioned plugins?) In that, we should probably copy prior-art here. Rather than re-discovering already solved problems?

Reply:
I know for certain that `npm` uses a temporary folder for `git clone` during the `npm install` process (but it also has other special behaviors around git based deps, that are not applicable to Helm).

I checked the `~/go/pkg` folder and it does not seem to have the `.git` folders there, so at least that implies that some sort of an intermediate method must have been used to perform the cloning?

A quick scan of pipenv code also implies a temp folder: https://github.com/pypa/pipenv/blob/da751e1f5568c72e013af9d977ad54c26e924d0e/pipenv/utils/dependencies.py#L749 (although they also have a special behavior for `git+file` which might be interesting to learn from).

---

If the question is about the overall approach - I think at least `npm` follows a very similar approach, from what I've seen - the end result after cloning does go through their `npm pack` (I think) process, because the final output in the `node_modules` folder looks the same way as if it had been `npm publish`ed to the registry.

---

I'm not familiar enough with the process and the implementation on the Helm side just yet, but it might make sense to do a `helm package` instead of treating it as if it's a `file://` dependency - not sure about any additional steps it performs. It seems the original implementation PR does it's own tarball build, not sure why just yet - still reviewing it.